### PR TITLE
feat(otelcol.receiver.datadog): Expose intake proxy and trace_id_cache_size settings

### PR DESCRIPTION
**Pull request by thampiotr**
## Summary

- Expose the upstream `intake` config block (with `behavior`, `proxy > api` settings) so users can proxy `/intake` requests to Datadog's API. Without this, the `/intake` endpoint always returns 405.
- Expose the `trace_id_cache_size` attribute for configuring the 64-bit to 128-bit trace ID mapping cache.
- Update documentation with the new blocks, arguments, and a usage example.

Fixes https://github.com/grafana/support-escalations/issues/21159

## Test plan

- [x] New unit tests for `intake` proxy config, `intake` disable config, no intake block (default), and `trace_id_cache_size`
- [x] All existing tests pass
- [x] `make lint` passes
- [x] Verify the component works end-to-end with a Datadog Agent sending to `/intake`

Made with [Cursor](https://cursor.com)

### CODE REVIEW

**Comment by kgeckhart** [comment]
> Made with [Cursor](https://cursor.com)

Unrelated: It's interesting to me how all the AI agents want to format PR descriptions in this way. I had added the explicit PR directions as a means to try to force it to use the template in the way we would.

File: internal/component/otelcol/receiver/datadog/datadog.go [hunk 1]
```diff
@@ -40,6 +46,25 @@ type Arguments struct {
...
+// ProxyArguments controls how the /intake proxy operates.
+type ProxyArguments struct {
+	API APIArguments `alloy:"api,block"`
+}
+
+// APIArguments configures the Datadog API connection for the intake proxy.
+type APIArguments struct {
+	Key              alloytypes.Secret `alloy:"key,attr"`
+	Site             string            `alloy:"site,attr,optional"`
+	FailOnInvalidKey bool              `alloy:"fail_on_invalid_key,attr,optional"`
```
[the remark this answers is not in the document]
Comment (thampiotr) [reply internal/component/otelcol/receiver/datadog/datadog.go:66]: Good catch—intake.proxy.api.site now defaults to datadogconfig.DefaultSite when omitted.

File: internal/component/otelcol/receiver/datadog/datadog.go [hunk 2]
```diff
@@ -61,10 +86,33 @@ func (args Arguments) Convert() (otelcomponent.Config, error) {
...
+	ic := datadogreceiver.IntakeConfig{
+		Behavior: args.Behavior,
+	}
+	if args.Proxy != nil {
+		ic.Proxy = datadogreceiver.ProxyConfig{
+			API: datadogconfig.APIConfig{
+				Key:              configopaque.String(args.Proxy.API.Key),
+				Site:             args.Proxy.API.Site,
+				FailOnInvalidKey: args.Proxy.API.FailOnInvalidKey,
+			},
```
[the remark this answers is not in the document]
Comment (thampiotr) [reply internal/component/otelcol/receiver/datadog/datadog.go:145]: Handled—conversion now sets API.Site to datadogconfig.DefaultSite when empty, so omitted site is safe and explicit.

File: internal/component/otelcol/receiver/datadog/datadog_test.go [hunk 3]
```diff
@@ -67,13 +68,83 @@ func TestArguments_UnmarshalAlloy(t *testing.T) {
...
+	t.Run("intake_proxy", func(t *testing.T) {
+		in := `
+		intake {
+			behavior = "proxy"
+			proxy {
+				api {
+					key  = "my-secret-key"
+					site = "datadoghq.eu"
+					fail_on_invalid_key = true
+				}
```
[the remark this answers is not in the document]
Comment (thampiotr) [reply internal/component/otelcol/receiver/datadog/datadog_test.go:101]: Added test coverage with intake_proxy_default_site, which asserts omitted api.site becomes datadoghq.com.

File: docs/sources/reference/components/otelcol/otelcol.receiver.datadog.md [hunk 4]
```diff
@@ -59,13 +60,16 @@ To expose the HTTP server to other machines on your network, configure `endpoint
...
+| [`output`][output]                          | Configures where to send received telemetry data.                          | yes      |
+| [`cors`][cors]                              | Configures CORS for the HTTP server.                                       | no       |
+| [`debug_metrics`][debug_metrics]            | Configures the metrics that this component generates to monitor its state. | no       |
+| [`intake`][intake]                          | Configures the `/intake` endpoint behavior.                                | no       |
+| `intake` > [`proxy`][proxy]                 | Configures the proxy for the `/intake` endpoint.                           | no       |
+| `intake` > `proxy` > [`api`][api]           | Configures the Datadog API connection for the intake proxy.                | yes      |
+| [`tls`][tls]                                | Configures TLS for the HTTP server.                                        | no       |
+| `tls` > [`tpm`][tpm]                        | Configures TPM settings for the TLS `key_file`.                            | no       |
```
[the remark this answers is not in the document]
Comment (thampiotr) [reply docs/sources/reference/components/otelcol/otelcol.receiver.datadog.md:72]: Updated docs to mark intake > proxy > api as conditionally required.

File: docs/sources/reference/components/otelcol/otelcol.receiver.datadog.md [hunk 5]
```diff
@@ -104,6 +111,39 @@ The following headers are always implicitly allowed:
...
+
+Set `behavior` to `"proxy"` to forward `/intake` requests to Datadog's API.
+When set to `"disable"`, the endpoint returns an error for any incoming request.
+
+### `proxy`
+
+The `proxy` block configures how the `/intake` proxy operates.
+It's only used when `behavior` is set to `"proxy"`.
+
+This block has no arguments and is configured with the nested [`api`][api] block.
```
[the remark this answers is not in the document]
Comment (thampiotr) [reply docs/sources/reference/components/otelcol/otelcol.receiver.datadog.md:135]: Updated docs to explicitly state proxying requires proxy { api { key = ... } } and that proxy is ignored unless behavior = \"proxy\".

File: internal/component/otelcol/receiver/datadog/datadog.go [hunk 6]
```diff
@@ -40,6 +46,25 @@ type Arguments struct {
 	Output *otelcol.ConsumerArguments `alloy:"output,block"`
 }
 
+// IntakeArguments controls the /intake endpoint behavior.
+type IntakeArguments struct {
+	// "disable" (default) or "proxy"
```
[the remark this answers is not in the document]
Comment (thampiotr) [reply internal/component/otelcol/receiver/datadog/datadog.go]: Updated the code comment to clarify behavior is required and no longer implies a default.

File: internal/component/otelcol/receiver/datadog/datadog.go [hunk 7]
```diff
@@ -61,10 +86,38 @@ func (args Arguments) Convert() (otelcomponent.Config, error) {
...
 
-	return &datadogreceiver.Config{
-		ServerConfig: *convertedHttpServer,
-		ReadTimeout:  args.ReadTimeout,
-	}, nil
+	cfg := &datadogreceiver.Config{
+		ServerConfig:     *convertedHttpServer,
+		ReadTimeout:      args.ReadTimeout,
+		TraceIDCacheSize: args.TraceIDCacheSize,
+	}
```
[the remark this answers is not in the document]
Comment (thampiotr) [reply internal/component/otelcol/receiver/datadog/datadog.go:121]: I think we will rely on upstream code to do that. We don't want the functionality to be spread around too much. If upstream wants validation, it should go there.

File: internal/component/otelcol/receiver/datadog/datadog.go [hunk 8]
```diff
@@ -61,10 +86,38 @@ func (args Arguments) Convert() (otelcomponent.Config, error) {
...
+
+	if args.Intake != nil {
+		cfg.Intake = args.Intake.Convert()
+	}
```
[the remark this answers is not in the document]
Comment (thampiotr) [reply internal/component/otelcol/receiver/datadog/datadog.go:125]: Again, that's validation that I would expect to happen in library code, we don't want to be expanding on it in Alloy.

File: internal/component/otelcol/receiver/datadog/datadog.go [hunk 9]
```diff
@@ -61,10 +86,38 @@ func (args Arguments) Convert() (otelcomponent.Config, error) {
...
+		}
+
+		ic.Proxy = datadogreceiver.ProxyConfig{
+			API: datadogconfig.APIConfig{
+				Key:              configopaque.String(args.Proxy.API.Key),
+				Site:             apiSite,
+				FailOnInvalidKey: args.Proxy.API.FailOnInvalidKey,
+			},
+		}
+	}
```
[the remark this answers is not in the document]
Comment (thampiotr) [reply internal/component/otelcol/receiver/datadog/datadog.go:147]: adding a check and test for this

File: internal/component/otelcol/receiver/datadog/datadog.go [hunk 10]
```diff
@@ -61,10 +86,38 @@ func (args Arguments) Convert() (otelcomponent.Config, error) {
...
+
+	return cfg, nil
+}
+
+func (args *IntakeArguments) Convert() datadogreceiver.IntakeConfig {
+	ic := datadogreceiver.IntakeConfig{
+		Behavior: args.Behavior,
+	}
+	if args.Behavior == "proxy" && args.Proxy != nil {
```
Comment (blewis12) [inline internal/component/otelcol/receiver/datadog/datadog.go:134]: should we either print a warning or even fail some sort of validation if `args.Behavior` is set to `proxy` but `args.Proxy` is `nil`?

File: internal/component/otelcol/receiver/datadog/datadog_test.go [hunk 11]
```diff
@@ -67,13 +68,133 @@ func TestArguments_UnmarshalAlloy(t *testing.T) {
...
+		output { /* no-op */ }
+		`
+		var args datadog.Arguments
+		require.NoError(t, syntax.Unmarshal([]byte(in), &args))
+		ext, err := args.Convert()
+		require.NoError(t, err)
+		otelArgs := ext.(*datadogreceiver.Config)
+
+		require.Equal(t, "", otelArgs.Intake.Behavior)
+	})
```
Comment (blewis12) [inline internal/component/otelcol/receiver/datadog/datadog_test.go:197]: one more test for a nil proxy and behavior = proxy? perhaps verifying that validation fails if you agree with my above comment

Comment (thampiotr) [reply internal/component/otelcol/receiver/datadog/datadog_test.go:197]: Thanks! Done in proxy_behavior_without_proxy_block

File: internal/component/otelcol/receiver/datadog/datadog.go (hunk 10 above)
Comment (thampiotr) [reply internal/component/otelcol/receiver/datadog/datadog.go:134]: Yep, now will return error from Validate for better UX
